File: langfuse-integration/scripts/should_trace.ts
export type TracePolicy = {
  enabled: boolean;
  allowProviders: Set<string>;
  denyProviders: Set<string>;
  allowModels: Set<string>;
  denyModels: Set<string>;
};

const csv = (value?: string) =>
  new Set((value || "").split(",").map((s) => s.trim()).filter(Boolean));

export function loadTracePolicy(env: NodeJS.ProcessEnv = process.env): TracePolicy {
  return {
    enabled: (env.LANGFUSE_ENABLE || "true").toLowerCase() !== "false",
    allowProviders: csv(env.LANGFUSE_ALLOW_PROVIDERS),
    denyProviders: csv(env.LANGFUSE_DENY_PROVIDERS),
    allowModels: csv(env.LANGFUSE_ALLOW_MODELS),
    denyModels: csv(env.LANGFUSE_DENY_MODELS),
  };
}

export function shouldTrace(provider: string, model: string, policy = loadTracePolicy()): boolean {
  if (!policy.enabled) return false;
  if (policy.denyProviders.has(provider) || policy.denyModels.has(model)) return false;
  if (policy.allowProviders.size > 0 && !policy.allowProviders.has(provider)) return false;
  if (policy.allowModels.size > 0 && !policy.allowModels.has(model)) return false;
  return true;
}

File: langfuse-integration/scripts/should_trace.py
import os


def _csv(name: str) -> set[str]:
    raw = os.getenv(name, "")
    return {x.strip() for x in raw.split(",") if x.strip()}


def should_trace(provider: str, model: str) -> bool:
    if os.getenv("LANGFUSE_ENABLE", "true").lower() == "false":
        return False

    allow_providers = _csv("LANGFUSE_ALLOW_PROVIDERS")
    deny_providers = _csv("LANGFUSE_DENY_PROVIDERS")
    allow_models = _csv("LANGFUSE_ALLOW_MODELS")
    deny_models = _csv("LANGFUSE_DENY_MODELS")

    if provider in deny_providers or model in deny_models:
        return False
    if allow_providers and provider not in allow_providers:
        return False
    if allow_models and model not in allow_models:
        return False
    return True
